README.rst
pyAesCrypt
===============
.. image:: https://travis-ci.com/marcobellaccini/pyAesCrypt.svg?branch=master
    :target: https://travis-ci.com/marcobellaccini/pyAesCrypt
.. image:: https://pepy.tech/badge/pyaescrypt
    :target: https://pepy.tech/project/pyaescrypt

About pyAesCrypt
--------------------------
pyAesCrypt is a Python 3 file-encryption module and script that uses AES256-CBC to encrypt/decrypt files and binary streams.

pyAesCrypt is compatible with the `AES Crypt`_ `file format`_ (version 2).

It is Free Software, released under the `Apache License, Version 2.0`_.

pyAesCrypt is brought to you by Marco Bellaccini - marco.bellaccini(at!)gmail.com.
 
IMPORTANT SECURITY NOTE: version 2 of the AES Crypt file format does not authenticate the "file size modulo 16" byte. This implies that an attacker  
with write access to the encrypted file may alter the corresponding plaintext file size by up to 15 bytes.

NOTE: there is no low-level memory management in Python, hence it is not possible to wipe memory areas were sensitive information was stored.

Module usage example
------------------------
Here is an example showing encryption and decryption of a file:

.. code:: python

    import pyAesCrypt
    password = "please-use-a-long-and-random-password"
    # encrypt
    pyAesCrypt.encryptFile("data.txt", "data.txt.aes", password)
    # decrypt
    pyAesCrypt.decryptFile("data.txt.aes", "dataout.txt", password)

**This is the most straightforward way to use pyAesCrypt, and should be preferred.**

If you need to specify a custom buffer size (default is 64KB), you can pass it as an optional argument:

.. code:: python

    import pyAesCrypt
    # custom encryption/decryption buffer size (default is 64KB)
    bufferSize = 128 * 1024
    password = "please-use-a-long-and-random-password"
    # encrypt
    pyAesCrypt.encryptFile("data.txt", "data.txt.aes", password, bufferSize)
    # decrypt
    pyAesCrypt.decryptFile("data.txt.aes", "dataout.txt", password, bufferSize)

In case you need it, you can work with binary streams too:

.. code:: python

    import pyAesCrypt
    from os import stat, remove
    # encryption/decryption buffer size - 64K
    # with stream-oriented functions, setting buffer size is mandatory
    bufferSize = 64 * 1024
    password = "please-use-a-long-and-random-password"
    
    # encrypt
    with open("data.txt", "rb") as fIn:
        with open("data.txt.aes", "wb") as fOut:
            pyAesCrypt.encryptStream(fIn, fOut, password, bufferSize)
    
    # get encrypted file size
    encFileSize = stat("data.txt.aes").st_size
    
    # decrypt
    with open("data.txt.aes", "rb") as fIn:
        try:
            with open("dataout.txt", "wb") as fOut:
                # decrypt file stream
                pyAesCrypt.decryptStream(fIn, fOut, password, bufferSize, encFileSize)
        except ValueError:
            # remove output file on error
            remove("dataout.txt")

you can also perform in-memory encryption/decryption (using BytesIO):

.. code:: python

    import pyAesCrypt
    import io
    
    bufferSize = 64 * 1024
    password = "please-use-a-long-and-random-password"
    
    # binary data to be encrypted
    pbdata = b"This is binary plaintext \x00\x01"
    
    # input plaintext binary stream
    fIn = io.BytesIO(pbdata)
    
    # initialize ciphertext binary stream
    fCiph = io.BytesIO()
    
    # initialize decrypted binary stream
    fDec = io.BytesIO()
    
    # encrypt stream
    pyAesCrypt.encryptStream(fIn, fCiph, password, bufferSize)
    
    # print encrypted data
    print("This is the ciphertext:\n" + str(fCiph.getvalue()))
    
    # get ciphertext length
    ctlen = len(fCiph.getvalue())
    
    # go back to the start of the ciphertext stream
    fCiph.seek(0)
    
    # decrypt stream
    pyAesCrypt.decryptStream(fCiph, fDec, password, bufferSize, ctlen)
    
    # print decrypted data
    print("Decrypted data:\n" + str(fDec.getvalue()))



Script usage examples
------------------------
Encrypt file test.txt in test.txt.aes:

	pyAesCrypt -e test.txt

Decrypt file test.txt.aes in test.txt:

	pyAesCrypt -d test.txt.aes
	
Encrypt file test.txt in test2.txt.aes:

	pyAesCrypt -e test.txt -o test2.txt.aes

Decrypt file test.txt.aes in test2.txt:

	pyAesCrypt -d test.txt.aes -o test2.txt

FAQs
------------------------
- *Is pyAesCrypt malware?*

  **NO!** Of course it isn't!

  Nevertheless, being a module, it can be used by any other software, including malware.
  
  In fact, it has been reported that it is used as crypto library by some ransomware.

.. _AES Crypt: https://www.aescrypt.com
.. _file format: https://www.aescrypt.com/aes_file_format.html
.. _Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0


HISTORY.rst
History
==================

6.0.0 (May 2021)
~~~~~~~~~~~~~~~~~~
* Updated password complexity check
* Set a default buffer size

5.0.0 (Gen 2021)
~~~~~~~~~~~~~~~~~~
* Fixed misleading messages when IOErrors occur (some public API exceptions where changed)
* Switched to Semantic Versioning

0.4.4 (Dec 2020)
~~~~~~~~~~~~~~~~~~
* Improved decryption speed
* Cross-platform CI

0.4.3 (Apr 2019)
~~~~~~~~~~~~~~~~~~
* Fixed bug when deleting bad output file (this was unnoticeable when running on GNU/Linux)

0.4.2 (Sep 2018)
~~~~~~~~~~~~~~~~~~
* Fixed bug when handling same input and output file 

0.4.1 (Sep 2018)
~~~~~~~~~~~~~~~~~~
* Preventing users from specifying the same input and output file 

0.4 (Aug 2018)
~~~~~~~~~~~~~~~~~~
* Support for binary stream encryption/decryption

0.3.1 (May 2018)
~~~~~~~~~~~~~~~~~~
* Fixed Python version compatibility classifier
* Removed useless stat

0.3 (Aug 2017)
~~~~~~~~~~~~~~~~~~
* Switched from `pycrypto`_ to `PyCA Cryptography`_ for crypto primitives
* Unittests clean-up

0.2.2 (Aug 2017)
~~~~~~~~~~~~~~~~~~
* Option to pass password as command-line argument to the script

0.2.1 (Feb 2017)
~~~~~~~~~~~~~~~~~~
* Better exceptions handling
* Code clean-up
* More unittests

0.2 (Jan 2017)
~~~~~~~~~~~~~~~~~~
* Modularized pyAesCrypt (and now the script calls the module for operations)
* Improved decryption speed (patch by Ben Fisher, see THANKS.rst)
* Improved encryption speed too
* Unittests
* Travis-CI integration
* Uploaded project to GitHub

0.1.2 (Jan 2016)
~~~~~~~~~~~~~~~~~~
* Bugfix

0.1 (Jan 2016)
~~~~~~~~~~~~~~~~~~
* First public release

.. _pycrypto: https://github.com/dlitz/pycrypto
.. _PyCA Cryptography: https://github.com/pyca/cryptography


pyAesCrypt/test_crypto.py
#==============================================================================
# Copyright 2020 Marco Bellaccini - marco.bellaccini[at!]gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#==============================================================================

# test suite for pyAesCrypt

import unittest
import os
import shutil
import filecmp
import subprocess
from os.path import isfile
import pyAesCrypt

# test file directory name
tfdirname = 'pyAesCryptTF'

# test file path prefix
prefix = tfdirname + '/'

# suffix for encrypted files
encsuffix = '.aes'

# suffix for decrypted files
decsuffix = '.decr'

# file names
filenames = [prefix+'empty', prefix+'block', prefix+'defbuffer',
             prefix+'small', prefix+'med', prefix+'tbuf']

# generate encrypted file names
encfilenames = list()
for f in filenames:
    encfilenames.append(f+encsuffix)

# generate decrypted file names
decfilenames = list()
for f in filenames:
    decfilenames.append(f+decsuffix)

# buffer size
bufferSize = 64 * 1024

# test password
password = "foopassword!1$A"


# function for generating test files
def genTestFiles():
    # empty file
    with open(filenames[0], 'wb') as fout:
        fout.write(b'')
    # one AES block (16 bytes) file
    with open(filenames[1], 'wb') as fout:
        fout.write(os.urandom(16))
    # default buffer size file
    with open(filenames[2], 'wb') as fout:
        fout.write(os.urandom(bufferSize))
    # small file
    with open(filenames[3], 'wb') as fout:
        fout.write(os.urandom(4))
    # med-sized file
    with open(filenames[4], 'wb') as fout:
        fout.write(os.urandom(2*bufferSize+19))
    # 3-buffer sized file
    with open(filenames[5], 'wb') as fout:
        fout.write(os.urandom(3*bufferSize))


# function for corrupting files
def corruptFile(fp, offset):
    # open file
    with open(fp, 'r+b') as ftc:
        # get to the byte to corrupt
        ftc.seek(offset)
        # read the byte
        rb = ftc.read(1)
        # back to the byte to corrupt
        ftc.seek(offset)
        # if byte is 'X', overwrite with 'Y'
        if rb == b'\x58':
            ftc.write(b'\x59')
        # else overwrite with 'X'
        else:
            ftc.write(b'\x58')


# test encryption and decryption
# NOTE: SOME TESTS REQUIRE AES CRYPT INSTALLED, WITH ITS BINARY IN $PATH
class TestEncDec(unittest.TestCase):
    # fixture for preparing the environment
    def setUp(self):
        # make directory for test files
        try:
            os.mkdir(tfdirname)
        # if directory exists, delete and re-create it
        except FileExistsError:
            # remove whole tree
            shutil.rmtree(tfdirname)
            os.mkdir(tfdirname)
        # generate test files
        genTestFiles()
        
    def tearDown(self):
        # delete test files
        for pt, ct, ou in zip(filenames, encfilenames, decfilenames):
            os.remove(pt)
            os.remove(ct)
            os.remove(ou)
        # delete directory for test files
        os.rmdir(tfdirname)
    
    # test pyAesCrypt encryption/decryption
    def test_enc_pyAesCrypt_dec_pyAesCrypt(self):
        for pt, ct, ou in zip(filenames, encfilenames, decfilenames):
            # encrypt file
            pyAesCrypt.encryptFile(pt, ct, password, bufferSize)
            # decrypt file
            pyAesCrypt.decryptFile(ct, ou, password, bufferSize)
            # check that the original file and the output file are equal
            self.assertTrue(filecmp.cmp(pt, ou))

    # test pyAesCrypt encryption/decryption with default buffer size
    def test_enc_pyAesCrypt_dec_pyAesCrypt_defbufsize(self):
        for pt, ct, ou in zip(filenames, encfilenames, decfilenames):
            # encrypt file
            pyAesCrypt.encryptFile(pt, ct, password)
            # decrypt file
            pyAesCrypt.decryptFile(ct, ou, password)
            # check that the original file and the output file are equal
            self.assertTrue(filecmp.cmp(pt, ou))
            
    # test encryption with pyAesCrypt and decryption with AES Crypt
    def test_enc_pyAesCrypt_dec_AesCrypt(self):
        for pt, ct, ou in zip(filenames, encfilenames, decfilenames):
            # encrypt file
            pyAesCrypt.encryptFile(pt, ct, password, bufferSize)
            # decrypt file
            subprocess.call(["aescrypt", "-d", "-p", password, "-o", ou, ct])
            # check that the original file and the output file are equal
            self.assertTrue(filecmp.cmp(pt, ou))
            
    # test encryption with AES Crypt and decryption with pyAesCrypt
    def test_enc_AesCrypt_dec_pyAesCrypt(self):
        for pt, ct, ou in zip(filenames, encfilenames, decfilenames):
            # encrypt file
            subprocess.call(["aescrypt", "-e", "-p", password, "-o", ct, pt])
            # decrypt file
            pyAesCrypt.decryptFile(ct, ou, password, bufferSize)
            # check that the original file and the output file are equal
            self.assertTrue(filecmp.cmp(pt, ou))

# test binary stream functions
class TestBS(unittest.TestCase):
    # fixture for preparing the environment
    def setUp(self):
        # make directory for test files
        try:
            os.mkdir(tfdirname)
        # if directory exists, delete and re-create it
        except FileExistsError:
            # remove whole tree
            shutil.rmtree(tfdirname)
            os.mkdir(tfdirname)
        # generate a test file
        with open(filenames[4], 'wb') as fout:
            fout.write(os.urandom(2*bufferSize+19))
        
    def tearDown(self):
        # delete test files
        os.remove(filenames[4])
        os.remove(encfilenames[4])
        os.remove(decfilenames[4])
        # delete directory for test files
        os.rmdir(tfdirname)
    
    # quick test for binary stream functions
    def test_bs_quick(self):
        # encrypt
        with open(filenames[4], "rb") as fIn:
            with open(encfilenames[4], "wb") as fOut:
                pyAesCrypt.encryptStream(fIn, fOut, password, bufferSize)
        
        # get encrypted file size
        encFileSize = os.stat(encfilenames[4]).st_size
        
        # decrypt
        with open(encfilenames[4], "rb") as fIn:
            with open(decfilenames[4], "wb") as fOut:
                # decrypt file stream
                pyAesCrypt.decryptStream(fIn, fOut, password, bufferSize,
                                         encFileSize)

        # check that the original file and the output file are equal
        self.assertTrue(filecmp.cmp(filenames[4], decfilenames[4]))


# test exceptions
class TestExceptions(unittest.TestCase):
    
    # test file path
    tfile = prefix + 'test.txt'
    
    # path of the copy of the test file
    tfilebak = tfile + '.bak'
    
    # fixture for preparing the environment
    def setUp(self):
        # make directory for test files
        try:
            os.mkdir(tfdirname)
        # if directory exists, delete and re-create it
        except FileExistsError:
            # remove whole tree
            shutil.rmtree(tfdirname)
            os.mkdir(tfdirname)
        # generate a test file
        with open(self.tfile, 'wb') as fout:
            fout.write(os.urandom(4))
        # copy of the test file
        shutil.copyfile(self.tfile, self.tfilebak)
        
    def tearDown(self):
        # remove whole directory tree
        shutil.rmtree(tfdirname)
    
    # test decryption with wrong password
    def test_dec_wrongpass(self):
        # encrypt file
        pyAesCrypt.encryptFile(self.tfile, self.tfile+'.aes', password,
                               bufferSize)
        # try to decrypt file using a wrong password
        # and check that ValueError is raised
        self.assertRaisesRegex(ValueError, ("Wrong password "
                                                "\(or file is corrupted\)."),
                               pyAesCrypt.decryptFile,
                               self.tfile + '.aes', self.tfile + '.decr',
                               'wrongpass', bufferSize)
                               
        # check that decrypted file was not created
        self.assertFalse(isfile(self.tfile + '.decr'))
            
    # test decryption of a non-AES-Crypt-format file
    def test_dec_not_AesCrypt_format(self):
        # encrypt file
        pyAesCrypt.encryptFile(self.tfile, self.tfile+'.aes', password,
                               bufferSize)
        # corrupt the 2nd byte (the 'E' of 'AES') - offset is 2-1=1
        corruptFile(self.tfile+'.aes', 1)
        
        # try to decrypt file
        # ...and check that ValueError is raised
        self.assertRaisesRegex(ValueError, ("File is corrupted or "
                                                "not an AES Crypt "
                                                "\(or pyAesCrypt\) file."),
                               pyAesCrypt.decryptFile,
                               self.tfile + '.aes', self.tfile + '.decr',
                               password, bufferSize)
        
        # check that decrypted file was not created
        self.assertFalse(isfile(self.tfile + '.decr'))
            
    # test decryption of an unsupported version of AES Crypt format
    def test_dec_unsupported_AesCrypt_format(self):
        # encrypt file
        pyAesCrypt.encryptFile(self.tfile, self.tfile+'.aes', password,
                               bufferSize)
        # corrupt the 4th byte
        corruptFile(self.tfile+'.aes', 3)
        
        # try to decrypt file
        # ...and check that ValueError is raised
        self.assertRaisesRegex(ValueError, ("pyAesCrypt is only "
                                                "compatible with version 2 of "
                                                "the AES Crypt file format."),
                               pyAesCrypt.decryptFile, self.tfile + '.aes',
                               self.tfile + '.decr', password, bufferSize)
                               
        # check that decrypted file was not created
        self.assertFalse(isfile(self.tfile + '.decr'))
            
    # test decryption of a file with bad hmac
    def test_dec_bad_hmac(self):
        # encrypt file
        pyAesCrypt.encryptFile(self.tfile, self.tfile+'.aes', password,
                               bufferSize)
                               
        # get file size
        fsize = os.stat(self.tfile+'.aes').st_size
        
        # corrupt hmac
        corruptFile(self.tfile+'.aes', fsize-1)
        
        # try to decrypt file
        # ...and check that ValueError is raised
        self.assertRaisesRegex(ValueError, ("Bad HMAC "
                                                "\(file is corrupted\)."),
                               pyAesCrypt.decryptFile, self.tfile + '.aes',
                               self.tfile + '.decr', password, bufferSize)
                               
        # check that decrypted file was deleted
        self.assertFalse(isfile(self.tfile + '.decr'))
            
    # test decryption of a truncated file (no complete hmac)
    def test_dec_trunc_file(self):
        # encrypt file
        pyAesCrypt.encryptFile(self.tfile, self.tfile+'.aes', password,
                               bufferSize)
                               
        # get file size
        fsize = os.stat(self.tfile+'.aes').st_size
        
        # truncate hmac (i.e.: truncate end of the file)
        with open(self.tfile+'.aes', 'r+b') as ftc:
            ftc.truncate(fsize-1)
        
        # try to decrypt file
        # ...and check that ValueError is raised
        self.assertRaisesRegex(ValueError, "File is corrupted.",
                               pyAesCrypt.decryptFile, self.tfile + '.aes',
                               self.tfile + '.decr', password, bufferSize)
                               
        # check that decrypted file was deleted
        self.assertFalse(isfile(self.tfile + '.decr'))
    
    # test same input and output file - encryption
    def test_samefile_enc(self):
        self.assertRaisesRegex(ValueError, ("Input and output files "
                                            "are the same."),
                               pyAesCrypt.encryptFile,
                               self.tfile, self.tfile,
                               'pass', bufferSize)
        # check that the original file was not modified
        self.assertTrue(filecmp.cmp(self.tfile, self.tfilebak))
    # test same input and output file - decryption
    def test_samefile_dec(self):
        self.assertRaisesRegex(ValueError, ("Input and output files "
                                            "are the same."),
                               pyAesCrypt.decryptFile,
                               self.tfile, self.tfile,
                               'pass', bufferSize)
        # check that the original file was not modified
        self.assertTrue(filecmp.cmp(self.tfile, self.tfilebak))
    
if __name__ == '__main__':
    unittest.main()


pyAesCrypt/crypto.py
#==============================================================================
# Copyright 2020 Marco Bellaccini - marco.bellaccini[at!]gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#==============================================================================

#==============================================================================
# pyAesCrypt
#
# pyAesCrypt is a Python file-encryption utility that uses AES256-CBC to
# encrypt/decrypt files and binary streams.
# pyAesCrypt is compatible with the AES Crypt (https://www.aescrypt.com/)
# file format (version 2).
# It uses PyCA Cryptography for crypto primitives and the operating system's
# random number generator (/dev/urandom on UNIX platforms, CryptGenRandom 
# on Windows).
#
# IMPORTANT SECURITY NOTE: version 2 of the AES Crypt file format does not
# authenticate the "file size modulo 16" byte. This implies that an attacker
# with write access to the encrypted file may alter the corresponding plaintext
# file size by up to 15 bytes.
#
# NOTE: there is no low-level memory management in Python, hence it is
# not possible to wipe memory areas were sensitive information was stored.
#==============================================================================

# pyAesCrypt module

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes, hmac
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from os import urandom
from os import stat, remove, path

# pyAesCrypt version - now semver
version = "6.0.0"

# default encryption/decryption buffer size - 64KB
bufferSizeDef = 64 * 1024

# maximum password length (number of chars)
maxPassLen = 1024

# AES block size in bytes
AESBlockSize = 16


# password stretching function
def stretch(passw, iv1):
    
    # hash the external iv and the password 8192 times
    digest = iv1 + (16 * b"\x00")
    
    for i in range(8192):
        passHash = hashes.Hash(hashes.SHA256(), backend=default_backend())
        passHash.update(digest)
        passHash.update(bytes(passw, "utf_16_le"))
        digest = passHash.finalize()
    
    return digest

# encrypt file function
# arguments:
# infile: plaintext file path
# outfile: ciphertext file path
# passw: encryption password
# bufferSize: optional buffer size, must be a multiple of
#             AES block size (16)
#             using a larger buffer speeds up things when dealing
#             with big files
#             Default is 64KB.
def encryptFile(infile, outfile, passw, bufferSize = bufferSizeDef):
    try:
        with open(infile, "rb") as fIn:
            # check that output file does not exist
            # or that, if exists, is not the same as the input file
            # (i.e.: overwrite if it seems safe)
            if path.isfile(outfile):
                if path.samefile(infile, outfile):
                    raise ValueError("Input and output files "
                                     "are the same.")
            try:
                with open(outfile, "wb") as fOut:
                    # encrypt file stream
                    encryptStream(fIn, fOut, passw, bufferSize)
                
            except IOError:
                raise ValueError("Unable to write output file.")
            
    except IOError:
        raise ValueError("Unable to read input file.")
                

# encrypt binary stream function
# arguments:
# fIn: input binary stream
# fOut: output binary stream
# passw: encryption password
# bufferSize: encryption buffer size, must be a multiple of
#             AES block size (16)
#             using a larger buffer speeds up things when dealing
#             with long streams
def encryptStream(fIn, fOut, passw, bufferSize):
    # validate bufferSize
    if bufferSize % AESBlockSize != 0:
        raise ValueError("Buffer size must be a multiple of AES block size.")
    
    if len(passw) > maxPassLen:
        raise ValueError("Password is too long.")

    # generate external iv (used to encrypt the main iv and the
    # encryption key)
    iv1 = urandom(AESBlockSize)
    
    # stretch password and iv
    key = stretch(passw, iv1)
    
    # generate random main iv
    iv0 = urandom(AESBlockSize)
    
    # generate random internal key
    intKey = urandom(32)
    
    # instantiate AES cipher
    cipher0 = Cipher(algorithms.AES(intKey), modes.CBC(iv0),
                     backend=default_backend())
    encryptor0 = cipher0.encryptor()
    
    # instantiate HMAC-SHA256 for the ciphertext
    hmac0 = hmac.HMAC(intKey, hashes.SHA256(),
                      backend=default_backend())
    
    # instantiate another AES cipher
    cipher1 = Cipher(algorithms.AES(key), modes.CBC(iv1),
                     backend=default_backend())
    encryptor1 = cipher1.encryptor()
    
    # encrypt main iv and key
    c_iv_key = encryptor1.update(iv0 + intKey) + encryptor1.finalize()
    
    # calculate HMAC-SHA256 of the encrypted iv and key
    hmac1 = hmac.HMAC(key, hashes.SHA256(),
                      backend=default_backend())
    hmac1.update(c_iv_key)

    # write header
    fOut.write(bytes("AES", "utf8"))
    
    # write version (AES Crypt version 2 file format -
    # see https://www.aescrypt.com/aes_file_format.html)
    fOut.write(b"\x02")
    
    # reserved byte (set to zero)
    fOut.write(b"\x00")
    
    # setup "CREATED-BY" extension
    cby = "pyAesCrypt " + version
    
    # write "CREATED-BY" extension length
    fOut.write(b"\x00" + bytes([1+len("CREATED_BY"+cby)]))
    
    # write "CREATED-BY" extension
    fOut.write(bytes("CREATED_BY", "utf8") + b"\x00" +
               bytes(cby, "utf8"))
    
    # write "container" extension length
    fOut.write(b"\x00\x80")
    
    # write "container" extension
    for i in range(128):
        fOut.write(b"\x00")
        
    # write end-of-extensions tag
    fOut.write(b"\x00\x00")
    
    # write the iv used to encrypt the main iv and the
    # encryption key
    fOut.write(iv1)
    
    # write encrypted main iv and key
    fOut.write(c_iv_key)
    
    # write HMAC-SHA256 of the encrypted iv and key
    fOut.write(hmac1.finalize())
    
    # encrypt file while reading it
    while True:
        # try to read bufferSize bytes
        fdata = fIn.read(bufferSize)
        
        # get the real number of bytes read
        bytesRead = len(fdata)
        
        # check if EOF was reached
        if bytesRead < bufferSize:
            # file size mod 16, lsb positions
            fs16 = bytes([bytesRead % AESBlockSize])
            # pad data (this is NOT PKCS#7!)
            # ...unless no bytes or a multiple of a block size
            # of bytes was read
            if bytesRead % AESBlockSize == 0:
                padLen = 0
            else:
                padLen = 16 - bytesRead % AESBlockSize
            fdata += bytes([padLen])*padLen
            # encrypt data
            cText = encryptor0.update(fdata) \
                    + encryptor0.finalize()
            # update HMAC
            hmac0.update(cText)
            # write encrypted file content
            fOut.write(cText)
            # break
            break
        # ...otherwise a full bufferSize was read
        else:
            # encrypt data
            cText = encryptor0.update(fdata)                                
            # update HMAC
            hmac0.update(cText)
            # write encrypted file content
            fOut.write(cText)
    
    # write plaintext file size mod 16 lsb positions
    fOut.write(fs16)
    
    # write HMAC-SHA256 of the encrypted file
    fOut.write(hmac0.finalize())

# decrypt file function
# arguments:
# infile: ciphertext file path
# outfile: plaintext file path
# passw: encryption password
# bufferSize: optional buffer size, must be a multiple of AES block size (16)
#             using a larger buffer speeds up things when dealing with
#             big files
#             Default is 64KB.
def decryptFile(infile, outfile, passw, bufferSize = bufferSizeDef):
    try:
        with open(infile, "rb") as fIn:
            # check that output file does not exist
            # or that, if exists, is not the same as the input file
            # (i.e.: overwrite if it seems safe)
            if path.isfile(outfile):
                if path.samefile(infile, outfile):
                    raise ValueError("Input and output files "
                                     "are the same.")
            try:
                with open(outfile, "wb") as fOut:
                    # get input file size
                    inputFileSize = stat(infile).st_size
                    try:
                        # decrypt file stream
                        decryptStream(fIn, fOut, passw, bufferSize,
                                      inputFileSize)
                    except ValueError as exd:
                        # should not remove output file here because it is still in use
                        # re-raise exception
                        raise ValueError(str(exd))
            
            except IOError:
                raise ValueError("Unable to write output file.")
            except ValueError as exd:
                # remove output file on error
                remove(outfile)
                # re-raise exception
                raise ValueError(str(exd))
                
    except IOError:
        raise ValueError("Unable to read input file.")
                    

# decrypt stream function
# arguments:
# fIn: input binary stream
# fOut: output binary stream
# passw: encryption password
# bufferSize: decryption buffer size, must be a multiple of AES block size (16)
#             using a larger buffer speeds up things when dealing with
#             long streams
# inputLength: input stream length
def decryptStream(fIn, fOut, passw, bufferSize, inputLength):
    # validate bufferSize
    if bufferSize % AESBlockSize != 0:
        raise ValueError("Buffer size must be a multiple of AES block size")
    
    if len(passw) > maxPassLen:
        raise ValueError("Password is too long.")

    fdata = fIn.read(3)
    # check if file is in AES Crypt format (also min length check)
    if (fdata != bytes("AES", "utf8") or inputLength < 136):
            raise ValueError("File is corrupted or not an AES Crypt "
                             "(or pyAesCrypt) file.")
        
    # check if file is in AES Crypt format, version 2
    # (the only one compatible with pyAesCrypt)
    fdata = fIn.read(1)
    if len(fdata) != 1:
        raise ValueError("File is corrupted.")
    
    if fdata != b"\x02":
        raise ValueError("pyAesCrypt is only compatible with version "
                         "2 of the AES Crypt file format.")
    
    # skip reserved byte
    fIn.read(1)
    
    # skip all the extensions
    while True:
        fdata = fIn.read(2)
        if len(fdata) != 2:
            raise ValueError("File is corrupted.")
        if fdata == b"\x00\x00":
            break
        fIn.read(int.from_bytes(fdata, byteorder="big"))
        
    # read external iv
    iv1 = fIn.read(16)
    if len(iv1) != 16:
        raise ValueError("File is corrupted.")
    
    # stretch password and iv
    key = stretch(passw, iv1)
    
    # read encrypted main iv and key
    c_iv_key = fIn.read(48)
    if len(c_iv_key) != 48:
        raise ValueError("File is corrupted.")
        
    # read HMAC-SHA256 of the encrypted iv and key
    hmac1 = fIn.read(32)
    if len(hmac1) != 32:
        raise ValueError("File is corrupted.")
    
    # compute actual HMAC-SHA256 of the encrypted iv and key
    hmac1Act = hmac.HMAC(key, hashes.SHA256(),
                         backend=default_backend())
    hmac1Act.update(c_iv_key)
    
    # HMAC check
    if hmac1 != hmac1Act.finalize():
        raise ValueError("Wrong password (or file is corrupted).")
    
    # instantiate AES cipher
    cipher1 = Cipher(algorithms.AES(key), modes.CBC(iv1),
                     backend=default_backend())
    decryptor1 = cipher1.decryptor()
    
    # decrypt main iv and key
    iv_key = decryptor1.update(c_iv_key) + decryptor1.finalize()
    
    # get internal iv and key
    iv0 = iv_key[:16]
    intKey = iv_key[16:]
    
    # instantiate another AES cipher
    cipher0 = Cipher(algorithms.AES(intKey), modes.CBC(iv0),
                     backend=default_backend())
    decryptor0 = cipher0.decryptor()
    
    # instantiate actual HMAC-SHA256 of the ciphertext
    hmac0Act = hmac.HMAC(intKey, hashes.SHA256(),
                         backend=default_backend())

    # decrypt ciphertext, until last block is reached
    while fIn.tell() < inputLength - 32 - 1 - AESBlockSize:
        # read data
        cText = fIn.read(
            min(
                bufferSize,
                inputLength - fIn.tell() - 32 - 1 - AESBlockSize
            )
        )
        # update HMAC
        hmac0Act.update(cText)
        # decrypt data and write it to output file
        fOut.write(decryptor0.update(cText))
        
    # last block reached, remove padding if needed
    
    # read last block
    
    # this is for empty files
    if fIn.tell() != inputLength - 32 - 1:
        cText = fIn.read(AESBlockSize)
        if len(cText) < AESBlockSize:
            raise ValueError("File is corrupted.")
    else:
        cText = bytes()
    
    # update HMAC
    hmac0Act.update(cText)
    
    # read plaintext file size mod 16 lsb positions
    fs16 = fIn.read(1)
    if len(fs16) != 1:
        raise ValueError("File is corrupted.")
    
    # decrypt last block
    pText = decryptor0.update(cText) + decryptor0.finalize()
    
    # remove padding
    toremove = ((16 - fs16[0]) % 16)
    if toremove != 0:
        pText = pText[:-toremove]
        
    # write decrypted data to output file
    fOut.write(pText)
    
    # read HMAC-SHA256 of the encrypted file
    hmac0 = fIn.read(32)
    if len(hmac0) != 32:
        raise ValueError("File is corrupted.")
    
    # HMAC check
    if hmac0 != hmac0Act.finalize():
        raise ValueError("Bad HMAC (file is corrupted).")   
